specs/icmp.rflx

```diff
@@ -45,93 +45,74 @@ package ICMP is
    type Uint_32_Unused is range 0 .. 0 with Size => 32;
    type Uint_24_Unused is range 0 .. 0 with Size => 24;
 
-   type Echo_Request_Reply_Message is
+   type ICMP_Message is
       message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
-         Identifier : Identifier;
-         Sequence_Number : Sequence_Number
-            then Data
-               with Length => Message'Last - Sequence_Number'Last;
-         Data : Opaque;
-      end message;
-
-   type Destination_Unreachable_Message is
-      message
-         Tag : Tag;
-         Code : Code_Destination_Unreachable;
-         Checksum : Checksum;
-         Unused : Uint_32_Unused
+         Tag : Tag
+            then Code_Destination_Unreachable
+               if Tag = Destination_Unreachable,
+            then Code_Time_Exceeded
+               if Tag = Time_Exceeded,
+            then Code_Redirect
+               if Tag = Redirect,
+            then Code_Zero
+               if (Tag = Echo_Reply or
+                  Tag = Echo_Request or
+                  Tag = Source_Quench or
+                  Tag = Parameter_Problem or
+                  Tag = Timestamp_Msg or
+                  Tag = Timestamp_Reply or
+                  Tag = Information_Request or
+                  Tag = Information_Reply);
+         Code_Destination_Unreachable : Code_Destination_Unreachable
+            then Checksum;
+         Code_Time_Exceeded : Code_Time_Exceeded
+            then Checksum;
+         Code_Redirect : Code_Redirect
+            then Checksum;
+         Code_Zero : Code_Zero
+            then Checksum;
```

Comment:
Every Message holds an 8 Bit Code Field. I think it could be better to make the Code-Field Opaque and check if the value-condition has been met on its outgoing edge. What do you think @treiher ?
```
Tag : Tag 
   then Code 
      with Length => 8
Code : Opaque
   then Checksum 
      if Tag = Destination_Unreachable and Code >= 0 and Code <= 5,
   then Checksum
      if ((Tag = Echo_Reply or Tag = Echo_Request  or Tag = Source_Quench ... ) and Code = 0),
                  
```

Reply:
> Every Message holds an 8 Bit Code Field. I think it could be better to make the Code-Field Opaque and check if the value-condition has been met on its outgoing edge.

Why `Opaque` instead of some dedicated range or modular type (or enum if discrete choices exist for `code`)? This would simplify the model further by removing the need for a `Length` attribute.

Reply:
Yes, a mod 8 integer type would be possible. There are discrete choices but they depend on the Tag e.g. if it is a `Destination Unreachable` Message, Code 0 means `Net_Unreachable` but if it is a `Time Exceeded` Msg, Code 0 means `TTL Exceeded`.

Reply:
`Opaque` cannot be compared to integers (at least it shouldn't be possible: #282). A separate scalar type for `Code` could be an option, but would probably introduce complications on the SPARK side (e.g. the need for explicit type conversions). In an offline discussion we decided to rather leave it as it is.